### PR TITLE
Prevent wrong deduplication in IncludeFile if vars_from is used (#66497)

### LINKED ISSUE #66497: include_role sometimes ignores dynamic vars_from parameter
##### SUMMARY

We're trying to implement roles, which can be individualized by supplying host related vars files. This doesn't seem to work under certain circumstances. Up to now we're not yet sure, whether this is a bug or works as designed.
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

include_role
##### ANSIBLE VERSION

```paste below
ansible 2.9.9
  config file = /adm/afroebel/ansible/unix/ansible.cfg
  configured module search path = [u'/adm/afroebel/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.16 (default, Oct 10 2019, 22:02:15) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
DEFAULT_GATHERING(/adm/afroebel/ansible/unix/ansible.cfg) = explicit
DEFAULT_MANAGED_STR(/adm/afroebel/ansible/unix/ansible.cfg) = {file}
DEFAULT_PRIVATE_ROLE_VARS(/adm/afroebel/ansible/unix/ansible.cfg) = True
PERSISTENT_CONNECT_TIMEOUT(/adm/afroebel/ansible/unix/ansible.cfg) = 30
```

##### OS / ENVIRONMENT

Ansible controller: Debian GNU/Linux 10
managed devices: Oracle Solaris 11.3/11.4

##### STEPS TO REPRODUCE

We've got 3 simple roles with exactly the same content, one task file and two vars files named like ansible hosts:

```yaml
# cksum test/roles/test_role/*/*/*
201185725 22 test/roles/test_role/solaris/tasks/main.yml
1196793317 19 test/roles/test_role/solaris/vars/test07
2721971076 19 test/roles/test_role/solaris/vars/test08
201185725 22 test/roles/test_role/test07/tasks/main.yml
1196793317 19 test/roles/test_role/test07/vars/test07
2721971076 19 test/roles/test_role/test07/vars/test08
201185725 22 test/roles/test_role/test08/tasks/main.yml
1196793317 19 test/roles/test_role/test08/vars/test07
2721971076 19 test/roles/test_role/test08/vars/test08

# cat test/roles/test_role/solaris/tasks/main.yml
---
- debug: var=var1

# cat test/roles/test_role/solaris/vars/test07
---
var1: "test07"

# cat test/roles/test_role/solaris/vars/test08
---
var1: "test08"
```
We now executed the playbook test/include_role.yml:
```yaml
---
- name:   "Test Playbook"
  hosts:  "test07:test08"
  tasks:
    - include_role:
        name:      "test_role/{{ inventory_hostname_short}}"
        vars_from: "{{ inventory_hostname_short }}"
```
It worked well and produced the following output:
```yaml
# ansible-playbook -i inventory test/include_role.yml

PLAY [Test Playbook] *******************************************************************

TASK [include_role : test_role/{{ inventory_hostname_short}}] **************************

TASK [test_role/test07 : debug] ********************************************************
ok: [test07] => {
    "var1": "test07"
}

TASK [test_role/test08 : debug] ********************************************************
ok: [test08] => {
    "var1": "test08"
}

PLAY RECAP *****************************************************************************
test07:  ok=1  changed=0  unreachable=0  failed=0  skipped=0  rescued=0  ignored=0
test08:  ok=1  changed=0  unreachable=0  failed=0  skipped=0  rescued=0  ignored=0
```
Tracing the ansible-playbook command using strace revealed, that tasks and vars files of either role were actually opened and read from.

The playbook test/include_role.yml was slightly changed in order to include a generic role test_role/solaris instead of the host related roles.
```yaml
---
- name:   "Test Playbook"
  hosts:  "test07:test08"
  tasks:
    - include_role:
        name:      "test_role/solaris"
        vars_from: "{{ inventory_hostname_short }}"
```

##### EXPECTED RESULTS

As the content of the generic role was exactly the same as the content of the host related roles, we would expect the playbook to produce the same output.

##### ACTUAL RESULTS

Even though the host related vars files contained different values for variable var1, ansible only used one of them during execution.

```yaml
# ansible-playbook -i inventory test/include_role.yml

PLAY [Test Playbook] *******************************************************************

TASK [include_role : test_role/solaris] ************************************************

TASK [test_role/solaris : debug] *******************************************************
ok: [test07] => {
    "var1": "test07"
}
ok: [test08] => {
    "var1": "test07"
}

PLAY RECAP *****************************************************************************
test07:  ok=1  changed=0  unreachable=0  failed=0  skipped=0  rescued=0  ignored=0
test08:  ok=1  changed=0  unreachable=0  failed=0  skipped=0  rescued=0  ignored=0
```
Tracing the ansible-playbook command using strace revealed, that this time only vars file test07 was opened and read from.

### PR DESCRIPTION
**Pull request by eqrx**
##### SUMMARY
This change modifies the `__eq__` method of `IncludedFile` to also expect the `_hosts` field to be equal to consider included files the same. 

Fixes #66497

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
IncludeFile

### CODE REVIEW

File: lib/ansible/playbook/included_file.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ def process_include_results(results, iterator, loader, variable_manager):
                         for from_arg in new_task.FROM_ARGS:
                             if from_arg in include_args:
                                 from_key = from_arg.replace('_from', '')
-                                new_task._from_files[from_key] = templar.template(include_args.pop(from_arg))
+                                new_task._from_files[from_key] = templar.template(include_args.get(from_arg))
```
Comment (bcoca) [inline lib/ansible/playbook/included_file.py]: this does not seem like the correct change, instead we need to make sure that the vars_from are loaded before pasisng to param hashing

Comment (sivel) [reply lib/ansible/playbook/included_file.py]: @bcoca those are 2 different things.  `role` deduplication, vs `IncludedFile` deduplication.

We "dedupe" `IncludedFile` due to many hosts including the same file, so we squash them into 1 object.  So it's a different mechanism.

And the means in which we dedupe or squash `IncludedFile` depends on `args` on in this case `include_args`.  So if we `.pop` them, then those args cannot be used for deduping.

Consider the example of individual hosts using a different `vars_from`.  Right now we would squash all of those hosts onto a single `IncludedFile` that didn't account for `vars_from`.

This seems like the right change to me.

Comment (bcoca) [reply lib/ansible/playbook/included_file.py]: but you dont have full args till you read the file. I becomes specially fun when the from itself is a template

**Comment by eqrx** [comment]
Hey, I have tested around a it and found this fix. I basically changed the __eq__ method of tasks to also check the hosts list. Could that break something that is not covered by tests?

**Comment by sivel** [comment]
> Could that break something that is not covered by tests?

I would have expected that to actually break existing tests.  We use the `__eq__` check to match whether results that came back from 2 hosts are the same, and then we squash the hosts onto a single `IncludedFile` object.

Thinking it through now, I think I see why it doesn't break anything, because they should never match. You would need to test a scenario like:

```
- hosts: localhost
  gather_facts: false
  tasks:
    - include_tasks: include_me.yml
      loop:
        - 1
        - 1
        - 1
```

And making sure it actually runs the tasks in `include_me.yml` 3 times.  I can test this more in the future, after I can get to a place where I'm ready to commit some code I'm in the middle of writing.

**Comment by eqrx** [comment]
@sivel that would be very nice of you. I can add your snippet to the testing but I am unsure what where to start investigating here

**Comment by eqrx** [comment]
> And making sure it actually runs the tasks in `include_me.yml` 3 times. I can test this more in the future, after I can get to a place where I'm ready to commit some code I'm in the middle of writing.

Can confirm that it gets run 3 times

**Pull request closed without merging**